File: src/main/java/net/slipp/service/MailService.java
package net.slipp.service;

import java.io.IOException;
import java.util.Map;

import javax.annotation.Resource;
import javax.mail.Message;
import javax.mail.MessagingException;
import javax.mail.internet.InternetAddress;
import javax.mail.internet.MimeMessage;

import net.slipp.domain.tag.Tag;
import net.slipp.domain.user.SocialUser;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.mail.javamail.JavaMailSender;
import org.springframework.mail.javamail.MimeMessagePreparator;
import org.springframework.scheduling.annotation.Async;
import org.springframework.stereotype.Service;
import org.springframework.ui.freemarker.FreeMarkerTemplateUtils;

import com.google.common.collect.Maps;

import freemarker.template.Configuration;
import freemarker.template.Template;

@Service
public class MailService {
    private static Logger log = LoggerFactory.getLogger(MailService.class);
    
    private static final String DEFAULT_ADMIN_EMAIL = "javajigi@gmail.com";

    @Resource(name = "mailSender")
    JavaMailSender mailSender;

    @Resource(name = "freemarkerConfiguration")
    Configuration configuration;

    @Async
    public void sendPasswordInformation(final SocialUser user) {
        MimeMessagePreparator preparator = new MimeMessagePreparator() {
            @Override
            public void prepare(MimeMessage mimeMessage) throws MessagingException {
                mimeMessage.setRecipient(Message.RecipientType.TO, new InternetAddress(user.getEmail()));
                mimeMessage.setFrom(new InternetAddress(DEFAULT_ADMIN_EMAIL));
                mimeMessage.setSubject("SLiPP에 오신것을 환영합니다!");

                String fileName = "passwordinfo.ftl";
                Map<String, Object> params = Maps.newHashMap();
                params.put("user", user);
                mimeMessage.setText(createMailTemplate(fileName, params), "utf-8", "html");
            }
        };
        mailSender.send(preparator);
    }
    
    @Async
    public void sendNewTagInformation(final Tag tag) {
        MimeMessagePreparator preparator = new MimeMessagePreparator() {
            @Override
            public void prepare(MimeMessage mimeMessage) throws MessagingException {
                mimeMessage.setRecipient(Message.RecipientType.TO, new InternetAddress(DEFAULT_ADMIN_EMAIL));
                mimeMessage.setFrom(new InternetAddress(DEFAULT_ADMIN_EMAIL));
                mimeMessage.setSubject("태그 추가 요청합니다.");

                String fileName = "requesttaginfo.ftl";
                Map<String, Object> params = Maps.newHashMap();
                params.put("tag", tag);
                mimeMessage.setText(createMailTemplate(fileName, params), "utf-8", "html");
            }
        };
        mailSender.send(preparator);        
    }

    private String createMailTemplate(String fileName, Map<String, Object> params) {
        Template template;
        try {
            template = configuration.getTemplate(fileName);
        } catch (IOException e) {
            log.error(e.toString());
            return null;
        }

        try {
            String result = FreeMarkerTemplateUtils.processTemplateIntoString(template, params);
            log.debug("mail auth body : {}", result);
            return result;
        } catch (Exception e) {
            log.error(e.toString());
            return null;
        }
    }
}

File: src/main/scala/net/slipp/service/MailService.scala
package net.slipp.service

import com.google.common.collect.Maps
import freemarker.template.Configuration
import freemarker.template.Template
import net.slipp.domain.tag.Tag
import net.slipp.domain.user.SocialUser
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.mail.javamail.JavaMailSender
import org.springframework.mail.javamail.MimeMessagePreparator
import org.springframework.scheduling.annotation.Async
import org.springframework.stereotype.Service
import org.springframework.ui.freemarker.FreeMarkerTemplateUtils
import javax.annotation.Resource
import javax.mail.Message
import javax.mail.MessagingException
import javax.mail.internet.InternetAddress
import javax.mail.internet.MimeMessage
import java.io.IOException
import java.util.Map

@Service object MailService {
  private val DEFAULT_ADMIN_EMAIL: String = "javajigi@gmail.com"
}

@Service class MailService {
  private val log: Logger = LoggerFactory.getLogger(classOf[MailService])
  @Resource(name = "mailSender") private[service] var mailSender: JavaMailSender = null
  @Resource(name = "freemarkerConfiguration") private[service] var configuration: Configuration = null

  @Async def sendPasswordInformation(user: SocialUser) {
    val preparator: MimeMessagePreparator = new MimeMessagePreparator() {
      @throws(classOf[MessagingException])
      def prepare(mimeMessage: MimeMessage) {
        mimeMessage.setRecipient(Message.RecipientType.TO, new InternetAddress(user.getEmail))
        mimeMessage.setFrom(new InternetAddress(MailService.DEFAULT_ADMIN_EMAIL))
        mimeMessage.setSubject("SLiPP에 오신것을 환영합니다!")
        val fileName: String = "passwordinfo.ftl"
        val params = Maps.newHashMap[String, AnyRef]
        params.put("user", user)
        mimeMessage.setText(createMailTemplate(fileName, params), "utf-8", "html")
      }
    }
    mailSender.send(preparator)
  }

  @Async def sendNewTagInformation(tag: Tag) {
    val preparator: MimeMessagePreparator = new MimeMessagePreparator() {
      @throws(classOf[MessagingException])
      def prepare(mimeMessage: MimeMessage) {
        mimeMessage.setRecipient(Message.RecipientType.TO, new InternetAddress(MailService.DEFAULT_ADMIN_EMAIL))
        mimeMessage.setFrom(new InternetAddress(MailService.DEFAULT_ADMIN_EMAIL))
        mimeMessage.setSubject("태그 추가 요청합니다.")
        val fileName: String = "requesttaginfo.ftl"
        val params = Maps.newHashMap[String, AnyRef]
        params.put("tag", tag)
        mimeMessage.setText(createMailTemplate(fileName, params), "utf-8", "html")
      }
    }
    mailSender.send(preparator)
  }

  private def createMailTemplate(fileName: String, params: Map[String, AnyRef]): String = {
    var template = configuration.getTemplate(fileName)
    val result = FreeMarkerTemplateUtils.processTemplateIntoString(template, params)
    log.debug("mail auth body : {}", result)
    result
  }
}
